src/hba.c

```diff
@@ -940,17 +966,50 @@ static bool match_inet6(const struct HBAAddress *haddress, PgAddr *addr)
 	       (src[2] & mask[2]) == base[2] && (src[3] & mask[3]) == base[3];
 }
 
+
 static bool address_match(const struct HBAAddress *haddress, PgAddr *addr)
 {
+	struct addrinfo *gai_result, *gai;
+	int ret;
+	bool found;
+
 	if (haddress->flags & ADDRESS_ALL)
 		return true;
-	switch (haddress->family) {
-	case AF_INET:
-		return match_inet4(haddress, addr);
-	case AF_INET6:
-		return match_inet6(haddress, addr);
-	default:
-		return false;
+	if (!haddress->hostname) {
+		/* match IP address */
+		switch (haddress->family) {
+		case AF_INET:
+			return match_inet4(haddress, addr);
+		case AF_INET6:
+			return match_inet6(haddress, addr);
+		default:
+			return false;
+		}
+	} else {
+		/* match DNS name */
+		ret = getaddrinfo(haddress->hostname, NULL, NULL, &gai_result);
```

Comment:
This seems to be doing only half of what postgres does to validate the hostname:

> If a host name is specified (anything that is not an IP address range or a special key word is treated as a host name), that name is compared with the result of a reverse name resolution of the client's IP address (e.g., reverse DNS lookup, if DNS is used). Host name comparisons are case insensitive. If there is a match, then a forward name resolution (e.g., forward DNS lookup) is performed on the host name to check whether any of the addresses it resolves to are equal to the client's IP address. If both directions match, then the entry is considered to match. 

Source: https://www.postgresql.org/docs/current/auth-pg-hba-conf.html

Your code is only doing the forward resolution of the hostname, not the reverse lookup of the IP address.